File: typescript/tests/loopback.test.ts
/**
 * Tests for the loopback redirect listener.
 *
 * Ports the Rust reference suite. These bind real sockets on 127.0.0.1 and
 * drive them with real requests, because the failures worth catching here are
 * exactly the ones a mocked server cannot have: a port that will not re-bind, a
 * favicon request mistaken for the redirect, a response that never flushes.
 */

import { describe, it, expect } from "vitest";
import { LoopbackListener } from "../src/loopback";

describe("LoopbackListener", () => {
  it("binds a loopback port and reports it", async () => {
    const l = await LoopbackListener.bind();
    expect(l.port).toBeGreaterThan(0);
    expect(l.redirectUri).toBe(`http://127.0.0.1:${l.port}/callback`);
    l.close();
  });

  it("uses the literal address rather than localhost", async () => {
    // RFC 8252, and it avoids IPv6-vs-IPv4 resolution surprises.
    const l = await LoopbackListener.bind();
    expect(l.redirectUri).toContain("127.0.0.1");
    expect(l.redirectUri).not.toContain("localhost");
    l.close();
  });

  it("re-binds the exact port and path of a saved URI", async () => {
    // A saved client id is bound to its redirect URI exactly, so a later run
    // has to come back on the same port.
    const first = await LoopbackListener.bindOn(0, "/cb");
    const uri = first.redirectUri;
    const port = first.port;
    first.close();
    // Give the OS a moment to release the port.
    await new Promise((r) => setTimeout(r, 50));

    const again = await LoopbackListener.bindFor(uri);
    expect(again.port).toBe(port);
    expect(again.redirectUri).toBe(uri);
    again.close();
  });

  it("fails loudly when the port is taken", async () => {
    const held = await LoopbackListener.bind();
    // Better a clear failure the caller can answer by re-registering than
    // authorizing against a URI the server will reject.
    await expect(
      LoopbackListener.bindFor(held.redirectUri),
    ).rejects.toThrowError(/cannot bind loopback port/);
    held.close();
  });

  it("rejects a redirect URI with no port", async () => {
    await expect(
      LoopbackListener.bindFor("https://example.com/callback"),
    ).rejects.toThrowError(/names no port/);
  });

  it("normalises a path without a leading slash", async () => {
    const l = await LoopbackListener.bindOn(0, "cb");
    expect(l.redirectUri.endsWith("/cb")).toBe(true);
    l.close();
  });

  it("captures code and state from the redirect", async () => {
    const l = await LoopbackListener.bind();
    const waiting = l.wait(5000);

    const res = await fetch(`${l.redirectUri}?code=the_code&state=the_state`);
    expect(res.status).toBe(200);
    // The user sees an outcome rather than a browser error.
    expect(await res.text()).toContain("Signed in");

    await expect(waiting).resolves.toEqual({
      code: "the_code",
      state: "the_state",
    });
  });

  it("ignores a favicon request and keeps waiting", async () => {
    const l = await LoopbackListener.bind();
    const waiting = l.wait(5000);

    // A browser asks for this unprompted; treating it as the redirect would
    // abort the flow.
    const favicon = await fetch(`http://127.0.0.1:${l.port}/favicon.ico`);
    expect(favicon.status).toBe(404);

    await fetch(`${l.redirectUri}?code=c2&state=s2`);
    await expect(waiting).resolves.toMatchObject({ code: "c2" });
  });

  it("surfaces a denial as a typed error", async () => {
    const l = await LoopbackListener.bind();
    // Attach the assertions before triggering the redirect: the rejection can
    // land during the fetch, and a promise that rejects with no handler yet is
    // an unhandled rejection even though we go on to assert on it.
    const waiting = l.wait(5000);
    const assertions = Promise.all([
      expect(waiting).rejects.toThrowError(
        expect.objectContaining({ kind: "denied" }),
      ),
      expect(waiting).rejects.toThrowError(/access_denied — User said no/),
    ]);

    await fetch(
      `${l.redirectUri}?error=access_denied&error_description=User%20said%20no`,
    );
    await assertions;
  });

  it("rejects a redirect carrying neither an error nor a code", async () => {
    const l = await LoopbackListener.bind();
    const waiting = l.wait(5000);
    const assertion = expect(waiting).rejects.toThrowError(
      /neither an error nor a code/,
    );

    const res = await fetch(l.redirectUri);
    expect(res.status).toBe(400);
    await assertion;
  });

  it("times out when no redirect arrives", async () => {
    const l = await LoopbackListener.bind();
    await expect(l.wait(80)).rejects.toThrowError(/timed out/);
  });

  it("decodes percent escapes in the code and state", async () => {
    // Codes and state values are opaque and routinely contain characters that
    // must survive a round trip through the query string.
    const l = await LoopbackListener.bind();
    const waiting = l.wait(5000);

    await fetch(`${l.redirectUri}?code=a%2Fb&state=x%20y`);
    await expect(waiting).resolves.toEqual({ code: "a/b", state: "x y" });
  });
});

File: ruby/test/loopback_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "socket"

# Tests for the loopback redirect listener.
#
# Ports the Rust reference suite. These bind real sockets on 127.0.0.1 and
# drive them with real requests, because the failures worth catching here are
# exactly the ones a mocked server cannot have: a port that will not re-bind, a
# favicon request mistaken for the redirect, a response that never flushes.
class LoopbackTest < Minitest::Test
  AC = DatagroutConduit::AuthCode
  Listener = DatagroutConduit::AuthCode::LoopbackListener

  def setup
    # These talk to real sockets, so WebMock must stay out of the way.
    WebMock.allow_net_connect!
  end

  # Issue a bare GET and return [status, body]. Raw TCP rather than Net::HTTP
  # so nothing intercepts it and the request line is exactly what a browser
  # would send.
  def get(port, target)
    socket = TCPSocket.new("127.0.0.1", port)
    socket.write("GET #{target} HTTP/1.1\r\nHost: 127.0.0.1:#{port}\r\n\r\n")
    raw = socket.read.to_s
    socket.close
    head, _, body = raw.partition("\r\n\r\n")
    [head.lines.first.to_s.split[1].to_i, body]
  end

  # Fire a request from another thread and return the listener's outcome.
  def redirect_to(listener, target)
    requester = Thread.new do
      # Give wait() a moment to reach the accept loop.
      sleep 0.05
      get(listener.port, target)
    end
    outcome = begin
      listener.wait(timeout: 5)
    rescue AC::Error => e
      e
    end
    [outcome, requester.value]
  end

  def test_binds_a_loopback_port_and_reports_it
    listener = Listener.bind
    assert_operator listener.port, :>, 0
    assert_equal "http://127.0.0.1:#{listener.port}/callback", listener.redirect_uri
  ensure
    listener&.close
  end

  def test_redirect_uri_uses_the_literal_address_not_localhost
    # RFC 8252, and it avoids IPv6-vs-IPv4 resolution surprises.
    listener = Listener.bind
    assert_includes listener.redirect_uri, "127.0.0.1"
    refute_includes listener.redirect_uri, "localhost"
  ensure
    listener&.close
  end

  def test_bind_for_reuses_the_exact_port_and_path_of_a_saved_uri
    # A saved client id is bound to its redirect URI exactly, so a later run
    # has to come back on the same port.
    first = Listener.bind_on(0, "/cb")
    uri = first.redirect_uri
    port = first.port
    first.close

    again = Listener.bind_for(uri)
    assert_equal port, again.port
    assert_equal uri, again.redirect_uri
  ensure
    again&.close
  end

  def test_bind_for_fails_loudly_when_the_port_is_taken
    held = Listener.bind
    # Better a clear failure the caller can answer by re-registering than
    # authorizing against a URI the server will reject.
    error = assert_raises(AC::HttpError) { Listener.bind_for(held.redirect_uri) }
    assert_includes error.message, "cannot bind loopback port"
  ensure
    held&.close
  end

  def test_bind_for_rejects_a_uri_with_no_port
    error = assert_raises(AC::HttpError) { Listener.bind_for("https://example.com/callback") }
    assert_includes error.message, "names no port"
  end

  def test_normalises_a_path_without_a_leading_slash
    listener = Listener.bind_on(0, "cb")
    assert listener.redirect_uri.end_with?("/cb")
  ensure
    listener&.close
  end

  def test_captures_code_and_state_from_the_redirect
    listener = Listener.bind
    outcome, response = redirect_to(listener, "/callback?code=the_code&state=the_state")

    status, body = response
    assert_equal 200, status
    # The user sees an outcome rather than a browser error.
    assert_includes body, "Signed in"

    assert_equal "the_code", outcome.code
    assert_equal "the_state", outcome.state
  end

  def test_ignores_a_favicon_request_and_keeps_waiting
    listener = Listener.bind
    port = listener.port

    requester = Thread.new do
      sleep 0.05
      # A browser asks for this unprompted; treating it as the redirect would
      # abort the flow.
      favicon = get(port, "/favicon.ico")
      sleep 0.05
      [favicon, get(port, "/callback?code=c2&state=s2")]
    end

    redirect = listener.wait(timeout: 5)
    favicon, = requester.value

    assert_equal 404, favicon.first
    assert_equal "c2", redirect.code
  end

  def test_surfaces_a_denial_as_a_typed_error
    listener = Listener.bind
    outcome, = redirect_to(
      listener, "/callback?error=access_denied&error_description=User%20said%20no"
    )

    assert_kind_of AC::DeniedError, outcome
    assert_equal :denied, outcome.kind
    assert_includes outcome.message, "access_denied — User said no"
  end

  def test_rejects_a_redirect_with_neither_an_error_nor_a_code
    listener = Listener.bind
    outcome, response = redirect_to(listener, "/callback")

    assert_equal 400, response.first
    assert_kind_of AC::DiscoveryError, outcome
    assert_includes outcome.message, "neither an error nor a code"
  end

  def test_times_out_when_no_redirect_arrives
    listener = Listener.bind
    error = assert_raises(AC::HttpError) { listener.wait(timeout: 0.2) }
    assert_includes error.message, "timed out"
  end

  def test_decodes_percent_escapes_in_the_code_and_state
    # Codes and state values are opaque and routinely contain characters that
    # must survive a round trip through the query string.
    listener = Listener.bind
    outcome, = redirect_to(listener, "/callback?code=a%2Fb&state=x%20y")

    assert_equal "a/b", outcome.code
    assert_equal "x y", outcome.state
  end

  def test_wait_stops_listening_afterwards
    # One-shot: the port is released once the redirect is captured, so a later
    # run can re-bind it.
    listener = Listener.bind
    port = listener.port
    redirect_to(listener, "/callback?code=c&state=s")

    again = Listener.bind_on(port, "/callback")
    assert_equal port, again.port
  ensure
    again&.close
  end

  def test_a_timeout_also_releases_the_port
    listener = Listener.bind
    port = listener.port
    assert_raises(AC::HttpError) { listener.wait(timeout: 0.2) }

    again = Listener.bind_on(port, "/callback")
    assert_equal port, again.port
  ensure
    again&.close
  end

  def test_close_is_safe_to_call_more_than_once
    listener = Listener.bind
    listener.close
    listener.close
  end
end
